### PR TITLE
Preliminary support for user-defined functions in RecalcEngine

### PR DESCRIPTION
**Pull request by MalekiRe**

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs [hunk 1]
```diff
@@ -169,6 +176,77 @@ public async Task<FormulaValue> EvalAsync(string expressionText, CancellationTok
...
+            {
+                record = record.Add(p);
+            }
+
+            var check = new CheckWrapper(this, definition.Body, record);
+
+            var func = new UserDefinedTexlFunction(definition.Name, definition.ReturnType, definition.Parameters, check);
+            if (_customFuncs.ContainsKey(definition.Name))
+            {
+                throw new InvalidOperationException($"function {definition.Name} is already defined");
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs:197]: nit: Function (capitalization)

File: src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngineWorker.cs [hunk 2]
```diff
@@ -68,6 +68,7 @@ private void RecalcWorker2(string name)
                 (var irnode, var ruleScopeSymbol) = IRTranslator.Translate(binding);
 
                 var scope = this;
+
```
Comment (bryancroteau-MSFT) [inline src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngineWorker.cs:71]: nit: No need for this change

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Review by bryancroteau-MSFT — APPROVED** [review]

**Pull request merged**